src/mjolnir/graphenhancer.cc

```diff
@@ -187,40 +187,327 @@ void UpdateSpeed(DirectedEdge& directededge, const uint32_t density, const uint3
+      if (!diredge->forward()) {
+        std::reverse(shape.begin(), shape.end());
+      }
+      uint32_t hdg = std::round(
+          PointLL::HeadingAlongPolyline(shape, GetOffsetForHeading(diredge->classification(),
+                                                                   diredge->use())));
+
+      // Convert to inbound heading
+      heading = ((hdg + 180) % 360);
+      break;
+    }
+  }
+
+  // Iterate through outbound edges and get turn degrees from the candidate
+  // edge onto outbound driveable edges.
+  diredge = tile->directededge(node->edge_index());
+  for (uint32_t i = 0; i < node->edge_count(); i++, diredge++) {
+    // Skip opposing directed edge and any edge that is not a road. Skip any
+    // edges that are not driveable outbound.
+    if (i == directededge.opp_local_idx() || !(diredge->forwardaccess() & kAutoAccess) ||
+        (directededge.restrictions() & (1 << diredge->localedgeidx())) != 0) {
+      continue;
+    }
+
+    // Get the heading of the outbound edge (unfortunately GraphEnhancer may
+    // not have yet computed and stored headings for this node).
+    auto shape = tile->edgeinfo(diredge->edgeinfo_offset()).shape();
+    if (!diredge->forward()) {
+      std::reverse(shape.begin(), shape.end());
+    }
+    uint32_t to_heading =
+        std::round(PointLL::HeadingAlongPolyline(shape, GetOffsetForHeading(diredge->classification(),
+                                                                            diredge->use())));
+    // Store outgoing turn type for any driveable edges
+    uint32_t turndegree = GetTurnDegree(heading, to_heading);
+    outgoing_turn_type.insert(Turn::GetType(turndegree));
+  }
+}
+
+// Update the rightmost lane as needed.
+void EnhanceRightLane(const DirectedEdge directededge,
+                      const uint32_t idx,
+                      NodeInfo& startnodeinfo,
+                      GraphTileBuilder& tilebuilder,
+                      GraphReader& reader,
+                      std::mutex& lock,
+                      std::vector<uint16_t>& enhanced_tls) {
+  std::set<Turn::Type> outgoing_turn_type;
+  GetTurnTypes(directededge, idx, outgoing_turn_type, startnodeinfo, tilebuilder, reader, lock);
+
+  size_t index = enhanced_tls.size() - 1;
+  uint16_t tl = enhanced_tls[index];
+
+  if (outgoing_turn_type.find(Turn::Type::kSlightRight) != outgoing_turn_type.end()) {
+    // assume the slight right is the through if kStraight is not found
+    if (tl == kTurnLaneThrough &&
+        outgoing_turn_type.find(Turn::Type::kStraight) != outgoing_turn_type.end()) {
+      tl |= kTurnLaneSlightRight;
+    }
+  }
+
+  if (outgoing_turn_type.find(Turn::Type::kRight) != outgoing_turn_type.end())
+    tl |= kTurnLaneRight;
+  if (outgoing_turn_type.find(Turn::Type::kSharpRight) != outgoing_turn_type.end())
+    tl |= kTurnLaneSharpRight;
+  enhanced_tls[index] = tl;
+}
+
+// Update the leftmost lane as needed.
+void EnhanceLeftLane(const DirectedEdge directededge,
```

Comment:
This is where we enhance the lanes by adding left lane information where the user added none.  We only add slight lefts when the through lanes have be accounted for in the turn types.